base/mpfr.jl

```diff
@@ -220,6 +220,12 @@ for (fJ, fC) in ((:si,:Clong), (:ui,:Culong))
     end
 end
 
+# Warning: the constants are MPFR implementation details from
+# `src/mpfr-impl.h`, search for `MPFR_EXP_ZERO`.
+mpfr_special_exponent(::Val{:zero}) = typemin(Clong) + true
+mpfr_special_exponent(::Val{:nan}) = mpfr_special_exponent(Val(:zero)) + true
+mpfr_special_exponent(::Val{:inf}) = mpfr_special_exponent(Val(:nan)) + true
```

Comment:
This chain of functions looks awfully like the pattern that is warned about in the docs on value types (https://docs.julialang.org/en/v1/manual/types/#%22Value-types%22), so I would suggest some performance benchmarks to make sure there isn't any problem with inference or compilation (especially with the recursive definitions from the inf case).

Also, why not just use the constant offsets from the MPFR header? This recursion just seems to overcomplicate the definitions when the header just does
```
# define MPFR_EXP_ZERO (MPFR_EXP_MIN+1)
# define MPFR_EXP_NAN  (MPFR_EXP_MIN+2)
# define MPFR_EXP_INF  (MPFR_EXP_MIN+3)
```

Reply:
```julia-repl
julia> @code_typed Base.MPFR.mpfr_special_exponent(Val(:inf))
CodeInfo(
1 ─     return -9223372036854775805
) => Int64
```

Maybe check such things yourself before cluttering PRs, if the result is not obvious to you? Just download the build artifact from the build checks below.

Reply:
What's the value of having these be functions over 3 different constants?

Reply:
Generally I prefer a function over many constants to keep the number of names that have to be imported from modules small. But this shouldn't be used outside of `Base.MPFR` anyway, so constants would be fine.

Should I change it to three different constants?